skglm/solvers/pn_solver.py

```diff
@@ -0,0 +1,295 @@
+import numpy as np
+from scipy.sparse import issparse
+from numba import njit
+
+from skglm.solvers.common import construct_grad, construct_grad_sparse
+
+
+def pn_solver(X, y, datafit, penalty, w_init=None, p0=10,
+              max_iter=20, max_epochs=1000, tol=1e-4, verbose=0):
+    """Run a Prox Newton solver.
+
+    Parameters
+    ----------
+    X : array, shape (n_samples, n_features)
+        Design matrix.
+
+    y : array, shape (n_samples,)
+        Target vector.
+
+    datafit : instance of BaseDatafit
+        Datafit object.
+
+    penalty : instance of BasePenalty
+        Penalty object.
+
+    w_init : array, shape (n_features,), default None
+        Initial value of coefficients.
+        If set to None, a zero vector is used instead.
+
+    p0 : int, default 10
+        Minimum number of groups to be included in the working set.
+
+    max_iter : int, default 20
+        Maximum number of iterations.
```

Comment:
this can be a bit mysterious to the user as there are many loops in the algorithm